pkg/testing/testwatch/testwatch.go

```diff
@@ -0,0 +1,145 @@
+// Copyright 2025 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package testwatch
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/klog/v2"
+	"kpt.dev/configsync/pkg/kinds"
+	"kpt.dev/configsync/pkg/util/log"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// WatchObject returns a new watcher for a typed resource.
+// The watcher will automatically be stopped when the Context is done.
+//
+// This is designed for simplified use in tests and is not likely to be suitable
+// for non-test usage.
+func WatchObject(ctx context.Context, c client.WithWatch, exampleObjList client.ObjectList) (watch.Interface, error) {
+	watcher, err := c.Watch(ctx, exampleObjList)
+	if err != nil {
+		return nil, err
+	}
+	go func() {
+		<-ctx.Done()
+		watcher.Stop()
```

Comment:
Is this needed given that the context is also passed to the watcher?

Reply:
Calling watcher.Stop allows for not just a downstream signal from caller to watcher but also allows waiting for an upstream signal from the watcher to the caller to confirm that the watcher is fully stopped.

As for whether the Watch method context actually stops the returned watcher... it depends on the watch impl. And frustratingly, there are a half dozen upstream wrappers that all behave subtly differently, especially fake ones used by tests. I spent months last year trying to get them fixed upstream but kept running into roadblocks from needing to be reverse compatible. So it's better to just be defensive.

Reply:
FWIW, this method isn't waiting for watcher.Stop to exit, so the return signal isn't actually being waited for here, but the caller should be waiting for the watcher.ResultChan to close, which has a similar effect.